models/gc_picai_baseline/utils/PicaiBaselineRunner.py

```diff
@@ -0,0 +1,67 @@
+"""
+---------------------------------------------------------
+Mhub / DIAG - Run Module for the PICAI baseline Algorithm
+---------------------------------------------------------
+
+---------------------------------------------------------
+Author: Sil van de Leemput
+Email:  sil.vandeleemput@radboudumc.nl
+---------------------------------------------------------
+"""
+
+import json
+from pathlib import Path
+
+from mhubio.core import Instance, InstanceData, IO, Module, ValueOutput, ClassOutput, Meta
+
+# Import the PICAI Classifier algorithm class from /opt/algorithm
+from process import csPCaAlgorithm as PicaiClassifier
+
+
+@ValueOutput.Name('prostate_cancer_likelihood')
+@ValueOutput.Label('ProstateCancerLikelihood')
+@ValueOutput.Type(float)
+@ValueOutput.Description('Likelihood of case-level prostate cancer.')
+class ProstateCancerLikelihood(ValueOutput):
+    pass
+
+
+class PicaiBaselineRunner(Module):
+
+    @IO.Instance()
+    @IO.Input('in_data_t2', 'mha:mod=mr:type=t2w', the='input T2 weighted prostate MR image')
+    @IO.Input('in_data_adc', 'mha:mod=mr:type=adc', the='input ADC prostate MR image')
+    @IO.Input('in_data_hbv', 'mha:mod=mr:type=hbv', the='input HBV prostate MR image')
+    @IO.Output('cancer_likelihood_json', 'cspca-case-level-likelihood.json', "json", bundle='model', the='output JSON file with PICAI baseline prostate cancer likelihood')
+    @IO.Output('cancer_detection_heatmap', 'cspca_detection_map.mha', "mha:mod=hm", bundle='model', the='output heatmap indicating prostate cancer likelihood')
+    @IO.OutputData('cancer_likelihood', ProstateCancerLikelihood, the='PICAI baseline prostate cancer likelihood')
+    def task(self, instance: Instance, in_data_t2: InstanceData, in_data_adc: InstanceData, in_data_hbv: InstanceData, cancer_likelihood_json: InstanceData, cancer_detection_heatmap: InstanceData, cancer_likelihood: ProstateCancerLikelihood) -> None:
+        # Initialize classifier object
+        classifier = PicaiClassifier()
+
+        # Specify input files (the order is important!)
+        classifier.scan_paths = [
+            Path(in_data_t2.abspath),
+            Path(in_data_adc.abspath),
+            Path(in_data_hbv.abspath),
+        ]
+
+        # Specify output files
+        classifier.cspca_detection_map_path = Path(cancer_detection_heatmap.abspath)
+        classifier.case_confidence_path = Path(cancer_likelihood_json.abspath)
+
+        # Run the classifier on the input images
+        classifier.process()
```

Comment:
As suspected: https://github.com/DIAGNijmegen/picai_nnunet_semi_supervised_gc_algorithm/blob/0b66c2fe224fca9045d94a77624b48450d787448/process.py#L251

We cannot capture any subprocess output that is not piped but flushes to the console out. I see several options, either you call the original pipeline via our subprocess wrapper, you call NNUnet using our subprocess wrapper from inisde your Runner Module, you wrap-up the code under /src and call that via cli. Any other suggestions from your side are welcome.

However, for consistency it is obviously highly important that no MHub implementation bypasses our output and logging.

Reply:
Yes, the nnU-Net inference is executed through the command line. If there is an option which keeps the algorithm still working on Grand Challenge, I can update my model inference pipeline such that the same model can run on GC and MHub. This would help maintainability of the code moving forward. 

Alternatively, the inference call could be made using pure Python, I think the nnU-Net framework does support that.

Who should pick this up?

Reply:
I like the minimally invasive solution, so I've implemented this change:
https://github.com/DIAGNijmegen/picai_nnunet_semi_supervised_gc_algorithm/commit/7235f7cfcdc27d24cde0693353abea697739bcd8

I verified if it still builds & runs the test case - all good!

Reply:
Adding a CLI wrapper is also possible, would this simply be a matter of adding the above script as cli.py? Let me know if it's useful to add this.

Reply:
Done!
https://github.com/DIAGNijmegen/picai_nnunet_semi_supervised_gc_algorithm/releases/tag/v2.1.2

I'll keep the CLI script for later then, because I don't have a test for it now.

Reply:
Awesome, thanks!

Reply:
I have updated it in the Dockerfile